### PR TITLE
Add Unit Tests to Shadow Freekicker Tactic

### LINKED ISSUE #819: Add Unit Tests For ShadowFreeKickerTactic
### Description of the task
See title.

Add getter functions to expose the Action parameters if needed.

### Acceptance criteria

- [ ] Add unit tests for the `ShadowEnemyFreeKicker` to improve both line and behavior coverage

### Blocked By
#773

### PR DESCRIPTION
**Pull request by HaomiaoZ**
### Description

Add unit tests to `Shadow Freekicker Tactic` when the enemy robot is at different locations
### Testing Done

Unit tests for `Shadow Freekicker Tactics` all passed
### Resolved Issues

resolves #819 
### Length Justification

Ascii graph contributes to around half of the lines, so the code is less than 500 lines

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/software/ai/hl/stp/tactic/shadow_freekicker_tactic_test.cpp
Comment (garethellis0) [inline src/software/ai/hl/stp/tactic/shadow_freekicker_tactic_test.cpp]: A quick comment at the start of the test explaining the scenario would be helpful.

Comment (garethellis0) [reply src/software/ai/hl/stp/tactic/shadow_freekicker_tactic_test.cpp]: A quick ascii diagram would potentially be very helpful. This website is very helpful for creating them: http://asciiflow.com/

Comment (garethellis0) [reply src/software/ai/hl/stp/tactic/shadow_freekicker_tactic_test.cpp]: Similar comment goes for the tests below as well.

**Review by garethellis0 — CHANGES_REQUESTED** [review]
It seems like there is a significant amount of duplication here. Could we make a test fixture and pull out some common functions / setup stuff? Not sure how much you know about test fixtures, but happy to explain more if need be!

In the interests of efficiency, going to hold off on an in-depth review of the tests themselves for now.

**Comment by HaomiaoZ** [comment]
@garethellis0 I am not too familiar with test fixture and it would be awesome if you can briefly explain or point me to any resource about it. I agree that setups for tests are repetitive and can be done better.

File: src/software/ai/hl/stp/tactic/shadow_freekicker_tactic_test.cpp
Comment (HaomiaoZ) [reply src/software/ai/hl/stp/tactic/shadow_freekicker_tactic_test.cpp]: ok, I will comment the code and consider adding ascii diagrams

**Comment by MathewMacDougall** [comment]
@HaomiaoZ here are some resources to get started:
* [Google's documentation](https://github.com/google/googletest/blob/master/googletest/docs/primer.md#test-fixtures-using-the-same-data-configuration-for-multiple-tests-same-data-multiple-tests)
* We use test fixtures already which should give some good examples. `src/software/backend/simulation/simulator_robot_test.cpp` has some good examples of generic helper functions, and `src/software/backend/simulation/physics/physics_robot_test.cpp` has a short example of the `SetUp` function.

Obviously we can also meet in person to help with this too, these are just good first places to look.

File: src/software/ai/hl/stp/tactic/shadow_freekicker_tactic_test.cpp
Comment (amrm3lm) [inline src/software/ai/hl/stp/tactic/shadow_freekicker_tactic_test.cpp]: could you clarify why we expect this to be around 0.62

Comment (HaomiaoZ) [reply src/software/ai/hl/stp/tactic/shadow_freekicker_tactic_test.cpp]: For sure, so this has some relation with our current implementation of the tactic. The `shadow freekicker tactic` would be assigned to two friendly robots to block the enemy's freekick. According to the rule, friendly robots cannot be closer than 0.5 m to the ball when an enemy is doing a freekick, so the distance between the ball with the line connecting two friendly robots is 0.5m with some room(1.05) and radius of the robot(0.09m) added to ensure that the edge of our robot won't be closer than 0.5 m to the ball during freekick.

I didn't find the reason why distance b is specified as 1.1 times of robot radius, but this is how `b` is calculated.

With `a` and `b` set up and known, the distance between the ball and friendly robot `c` then become 0.62292, which is close to 0.62.

image

File: src/software/ai/hl/stp/tactic/shadow_freekicker_tactic_test.cpp
Comment (kelvinkoon) [inline src/software/ai/hl/stp/tactic/shadow_freekicker_tactic_test.cpp]: In the same line as the 0.62 comment, it might be beneficial to use constants/defines to denote what 0.09, 0.01, 0.62, etc. mean (as opposed to having magic numbers).

**Review by kelvinkoon — COMMENTED** [review]
Suggesting changing to use defines/consts for the values or be more explicit in comments about what each number in the tests mean.

File: src/software/ai/hl/stp/tactic/shadow_freekicker_tactic_test.cpp
Comment (garethellis0) [reply src/software/ai/hl/stp/tactic/shadow_freekicker_tactic_test.cpp]: We should at least add a comment saying essentially this, just so that anyone looking at the code in the future can figure it out.

**Comment by garethellis0** [comment]
> @garethellis0 I am not too familiar with test fixture and it would be awesome if you can briefly explain or point me to any resource about it. I agree that setups for tests are repetitive and can be done better.

Sorry for the delay on this. Checkout https://github.com/UBC-Thunderbots/Software/blob/master/src/software/world/team_test.cpp for an example of how we use test fixtures. The general strategy is that duplicate work can be placed in `SetUp` and that will be run before each test. You can always define `TearDown` to run after each test, but I don't think we need that here.

File: src/software/ai/hl/stp/tactic/shadow_freekicker_tactic_test.cpp
Comment (amrm3lm) [inline src/software/ai/hl/stp/tactic/shadow_freekicker_tactic_test.cpp]: minor thing but since this is a constant i think it should have the 'const' keyword and the name should be capitalized, same thing for the other constants

Comment (HaomiaoZ) [reply src/software/ai/hl/stp/tactic/shadow_freekicker_tactic_test.cpp]: That's really good suggestion, I will do that

**Review by HaomiaoZ — COMMENTED** [review]
> > @garethellis0 I am not too familiar with test fixture and it would be awesome if you can briefly explain or point me to any resource about it. I agree that setups for tests are repetitive and can be done better.
> 
> Sorry for the delay on this. Checkout https://github.com/UBC-Thunderbots/Software/blob/master/src/software/world/team_test.cpp for an example of how we use test fixtures. The general strategy is that duplicate work can be placed in `SetUp` and that will be run before each test. You can always define `TearDown` to run after each test, but I don't think we need that here.

@garethellis0  I need to pass some parameter there so I created a function that takes parameters rather than `SetUp` as it seems that `SetUp` does not take any arguments

File: src/software/ai/hl/stp/tactic/shadow_freekicker_tactic_test.cpp
Comment (HaomiaoZ) [reply src/software/ai/hl/stp/tactic/shadow_freekicker_tactic_test.cpp]: Ascii graph added

File: src/software/ai/hl/stp/tactic/shadow_freekicker_tactic_test.cpp
Comment (HaomiaoZ) [reply src/software/ai/hl/stp/tactic/shadow_freekicker_tactic_test.cpp]: this information is added as an ascii graph

File: src/software/ai/hl/stp/tactic/shadow_freekicker_tactic_test.cpp
Comment (HaomiaoZ) [reply src/software/ai/hl/stp/tactic/shadow_freekicker_tactic_test.cpp]: Those constants are defined and commented in the test fixture

File: src/software/ai/hl/stp/tactic/shadow_freekicker_tactic_test.cpp
Comment (HaomiaoZ) [reply src/software/ai/hl/stp/tactic/shadow_freekicker_tactic_test.cpp]: All constants are capitalized and `const` keyword is added

**Comment by HaomiaoZ** [comment]
Ready for review this time

File: src/software/ai/hl/stp/tactic/shadow_freekicker_tactic_test.cpp
Comment (kelvinkoon) [inline src/software/ai/hl/stp/tactic/shadow_freekicker_tactic_test.cpp:866]: Files should have EOF extra line.

**Review by kelvinkoon — CHANGES_REQUESTED** [review]
ASCII art and changes seem to be implemented :+1: 
LGTM otherwise.

File: src/software/ai/hl/stp/tactic/shadow_freekicker_tactic_test.cpp
Comment (HaomiaoZ) [reply src/software/ai/hl/stp/tactic/shadow_freekicker_tactic_test.cpp:866]: I do have an extra line in CLion for some reason, but I added another line and now github can show the EOF line

File: src/software/ai/hl/stp/tactic/shadow_freekicker_tactic_test.cpp
Comment (HaomiaoZ) [reply src/software/ai/hl/stp/tactic/shadow_freekicker_tactic_test.cpp]: That is really a good catch! changed

**Review by garethellis0 — DISMISSED** [review]
LGTM pending other comments here

**Review by kelvinkoon — DISMISSED** [review]

**Comment by amrm3lm** [comment]
LGTM once travis passes

**Comment by HaomiaoZ** [comment]
Unfortunately Tarvis failed. Tried to build locally, software seems fine, but fail to build firmware. Is that the reason why Travis failed? @garethellis0

**Comment by garethellis0** [comment]
Seems like there were some issues with `EPSILON` not being defined, probably because we changed how we handle epsilon in https://github.com/UBC-Thunderbots/Software/pull/1196. Merging `master` into your branch _might_ help.

```

Use --sandbox_debug to see verbose messages from the sandbox
[32m[978 / 1,394][0m .../output/radio:radio_output; 5s remote-cache, linux-sandbox

[1A[Ksoftware/world/field.cpp: In member function 'bool Field::isFieldValid() const':
software/world/field.cpp:135:41: error: 'EPSILON' is not a member of 'GeomConstants'
     if (totalXLength() < GeomConstants::EPSILON ||
                                         ^~~~~~~
software/world/field.cpp:135:41: note: suggested alternative: 'FIXED_EPSILON'
     if (totalXLength() < GeomConstants::EPSILON ||
                                         ^~~~~~~
                                         FIXED_EPSILON
software/world/field.cpp:136:41: error: 'EPSILON' is not a member of 'GeomConstants'
         totalYLength() < GeomConstants::EPSILON)
                                         ^~~~~~~
software/world/field.cpp:136:41: note: suggested alternative: 'FIXED_EPSILON'
         totalYLength() < GeomConstants::EPSILON)
                                         ^~~~~~~
                                         FIXED_EPSILON
At global scope:
cc1plus: warning: unrecognized command line option '-Wno-unused-command-line-argument'
[32m[978 / 1,394][0m .../output/radio:radio_output; 5s remote-cache, linux-sandbox
```

Please take a look at the full travis log, it shows the exact commands that run and what failed (or check `.travis.yml` for the commands we run).

**Comment by amrm3lm** [comment]
> Seems like there were some issues with `EPSILON` not being defined, probably because we changed how we handle epsilon in #1196. Merging `master` into your branch _might_ help.
> 
> ```
> 
> Use --sandbox_debug to see verbose messages from the sandbox
> �[32m[978 / 1,394]�[0m .../output/radio:radio_output; 5s remote-cache, linux-sandbox
> 
> �[1A�[Ksoftware/world/field.cpp: In member function 'bool Field::isFieldValid() const':
> software/world/field.cpp:135:41: error: 'EPSILON' is not a member of 'GeomConstants'
>      if (totalXLength() < GeomConstants::EPSILON ||
>                                          ^~~~~~~
> software/world/field.cpp:135:41: note: suggested alternative: 'FIXED_EPSILON'
>      if (totalXLength() < GeomConstants::EPSILON ||
>                                          ^~~~~~~
>                                          FIXED_EPSILON
> software/world/field.cpp:136:41: error: 'EPSILON' is not a member of 'GeomConstants'
>          totalYLength() < GeomConstants::EPSILON)
>                                          ^~~~~~~
> software/world/field.cpp:136:41: note: suggested alternative: 'FIXED_EPSILON'
>          totalYLength() < GeomConstants::EPSILON)
>                                          ^~~~~~~
>                                          FIXED_EPSILON
> At global scope:
> cc1plus: warning: unrecognized command line option '-Wno-unused-command-line-argument'
> �[32m[978 / 1,394]�[0m .../output/radio:radio_output; 5s remote-cache, linux-sandbox
> ```
> 
> Please take a look at the full travis log, it shows the exact commands that run and what failed (or check `.travis.yml` for the commands we run).

interestingly i got the same error only after pulling from master

**Comment by garethellis0** [comment]
Try pulling `master` again, I think I just fixed this and merged to `master`.

**Comment by HaomiaoZ** [comment]
Thanks @garethellis0! Travis is fine now

**Review by garethellis0 — APPROVED** [review]

**Review by MathewMacDougall — APPROVED** [review]

**Pull request merged**
